hw/ip/prim/util/primgen.py

```diff
@@ -0,0 +1,383 @@
+
+    techlib_enums = []
+
+    # Insert the required generic library first to ensure it gets enum value 0
+    techlib_enums.append(_enum_name_for_techlib('generic', qualified=False))
+
+    for techlib in techlibs:
+        if techlib == 'generic':
+            # The generic implementation is required and handled separately.
+            continue
+        techlib_enums.append(_enum_name_for_techlib(techlib, qualified=False))
+
+    # Render prim_pkg.sv file
+    print("Creating prim_pkg.sv")
+    prim_pkg_sv_tpl_filepath = os.path.join(os.path.dirname(__file__),
+                                            'primgen', 'prim_pkg.sv.tpl')
+    prim_pkg_sv_tpl = Template(filename=prim_pkg_sv_tpl_filepath)
+
+    prim_pkg_sv = prim_pkg_sv_tpl.render(encoding="utf-8",
+                                         techlib_enums=techlib_enums)
+    with open('prim_pkg.sv', 'w') as f:
+        f.write(prim_pkg_sv)
+
+    # Copy prim_pkg.core (no changes needed)
+    prim_pkg_core_src = os.path.join(os.path.dirname(__file__), 'primgen',
+                                     'prim_pkg.core.tpl')
+    prim_pkg_core_dest = 'prim_pkg.core'
+    shutil.copyfile(prim_pkg_core_src, prim_pkg_core_dest)
+    print("Core file written to %s." % (prim_pkg_core_dest, ))
+
+
+def _instance_sv(prim_name, techlib, parameters):
+    s = "prim_{techlib}_{prim_name} #(\n"
+    s += ", ".join(".{p}({p})".format(p=p) for p in parameters)
+    s += ") u_impl_{techlib} (\n" \
+         "  .*\n" \
+         ");\n"
+    return s.format(prim_name=prim_name, techlib=techlib)
+
+
+def _create_instances(prim_name, techlibs, parameters):
+    """ Build SystemVerilog code instantiating primitives from the techlib """
+    techlibs_wo_generic = [
+        techlib for techlib in techlibs if techlib != 'generic'
+    ]
+    techlibs_generic_last = techlibs_wo_generic + ['generic']
+
+    if not techlibs_wo_generic:
+        # Don't output the if/else blocks if there no alternatives exist.
+        # We still want the generate block to keep hierarchical path names
+        # the stable, even if more than one techlib is found.
+        s = "generate begin : gen_generic\n"
+        s += _instance_sv(prim_name, "generic", parameters) + "\n"
+        s += "end endgenerate"
+        return s
+
+    nr_techlibs = len(techlibs_generic_last)
+    out = ""
+    for pos, techlib in enumerate(techlibs_generic_last):
+        is_first = pos == 0
+        is_last = pos == nr_techlibs - 1
+
+        s = ""
+        if not is_first:
+            s += "else "
+        if not is_last:
+            s += "if (Impl == {techlib_enum}) "
+
+        # TODO: wildcard port lists are against our style guide, but it's safer
+        # to let the synthesis tool figure out the connectivity than us trying
+        # to parse the port list into individual signals.
+        s += "begin : gen_{techlib}\n" + _instance_sv(prim_name, techlib,
+                                                      parameters) + "end "
```

Comment:
I think this should be ok for the time being. 

I recall that I wrote a simplistic parser for extracting parameters and port names for bootstrapping the boilerplate code for FPV testbenches (`util/fpvgen.py` and `util/fpvgen/sv_parse.py`). but that is super hacky and only works for the most common cases (i.e. no unpacked dimensions). and its not nearly as elegant as your regex solution above :).

Reply:
Ah, I didn't see that one before. I've opened https://github.com/lowRISC/opentitan/issues/1720 to keep track of that, and maybe one day we'll find a nice parser to use instead of our hacks.